### DISCUSSION TITLE
Multiple dynamic routes files at same path

### DISCUSSION THREAD

**Discussion by vikaskhunteta** [vercel/next.js#21595] [Help]
Hi

Is this possible to have multiple dynamic routes files under same path?

For e.g. in the root of `pages` folder can we have files like `[post].js`, `[page].js`, `[category].js` not like `pages/post/[post].js` or `pages/category/[category].js`

this comes into my mind because I have been rebuilding a WP site into React using Nextjs, we have URLs of posts, pages and categories which start from the root, I mean no classification like `/page/url` or `/post/url`  or `/category/url`

When I put multiple dynamic routes files in the root of `pages` folder it gives error `Failed to fetch API`  because both dynamic routes called at once while I used their `href` separately for posts I used `[post].js` and for category `[category].js`

So basically I want my dynamic routes same as WP permalink structure

Let me know if its possible to achieve or not

**Comment by mkrtchian** [comment]
Hello,

If your URL is dynamic from the root, like `/my-specific-page` or `/my-specific-post`, how can you know if "my-specific-page" is a page and not a post ? Do you have a pattern or so ?

Maybe you can have a single page handling all the dynamic routing of root level, and do the logic there to choose what you will fetch and display.

**Reply by vikaskhunteta** [reply]
Hi,

the `pages` or `posts` or `categories` will be decided by the `href` attribute path name which would have dynamic route like below

`pages` path name `/[page].js`, `posts` path name `/[post].js` and `categories` path name `/[category].js`

as the requests goes without reloading the page, why the request doesn't goes to their respective paths?

Do you think its possible without doing some logic inside the one of the root file?

**Reply by vikaskhunteta** [reply]
Hi,

Do you have any idea on how I can accomplish this? I want my all posts, pages and categories path to start from root.

I want to do this because we have same permalink structure for our current WP site and I want to convert that site into React

So when the old URL will hit the site I don't want them to redirect to new URL

**Reply by levranger** [reply]
Are there any updates?) I am facing the same problem(

**Reply by vikaskhunteta** [reply]
Seems to be no solution of this problem at this time, you may be want to try expressjs or some other nodejs server, I don't know whether it will help or not

**Reply by zkrony** [reply]
I am also facing this problem. Life is not easy as they think

**Reply by levranger** [reply]
> I am also facing this problem. Life is not easy as they think

I've solved.
You just need to handle all these routes in one file (use Url) 
Example if url is /team-123, we return component called team and pass there an id (123), and in component team we make a request to get the data according to id

**Reply by popeating** [reply]
i solved this way

starting from this situation:
https://mysite.com/category/[pagenumber] -> archive page for category
https://mysite.com/category/subcategory/[pagenumber] -> archive for subcategory (luckly im limited to 1 subcategory)
https://mysite.com/category/article -> single article for category
https://mysite.com/category/subcategory/article -> single article for subcategory

i setup the following structure:
[category]/[[...slug]].js

so that it matches /category /category/article /category/page/1 /category/subcategory and so on

--- only drawback is that the subcategory is part of the slug parameter, i tried with [category]/[[...subcategory]]/[[...slug]].js but i wasnt able to further split subcategory and slug ---

then in [[...slug]].js i do most of my logic, based on the category and slug i can see if im on an archive page or in a single post page and have the getStaticProps act accordingly, exectute the right query and pass the result to the slug component (along with a type of page i want to show), and using the type i can conditionally render the correct component that i created outside (Archive, SinglePost and so on).

Another couple of things i do, i generate with getStaticPaths() the newest 50 articles (with fallback: 'blocking') and the first pages of all categories archive (few people use/browse archive) <- let me know if this strategy can make sense!

Still working on it to make a solid and fleible base for all my future projects
Hope it helps

**Comment by popeating** [comment]
im stuck on same situation

#24593

**Comment by PrimozRome** [comment]
No solution here? This should be highly requested feature.

**Comment by 2stash** [comment]
This has been solved and the answer is linked in @popeating posts. 

https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes

At the root of pages you add a file called [...slug].js.  Name doesn't matter but the "..." tells nextjs to accept all routes.

So in the [...slug].js file you can check if the path is "/my-specific-page" or "/post/2021/10/5" and perform logic. If you wanted separate components or files, then a solution would be to have your logic in the [...slug].js file load normal React components based on the route parameters.

And you can get the route params by importing useRouter and get the query params.

https://nextjs.org/docs/api-reference/next/router

**Comment by hades200082** [comment]
To me it seems that NextJS should be able to tell whether a route results in a page or not.

The desired outcome here is that we can separate our page logic following the normal NextJS routing patterns while allowing different dynamic routes at the same level.

```text
app
   |- [product-slug]
   |- [category-slug]
   /* snip */
```

Why can't NextJS tell if one of those routes doesn't have any data and build accordingly?

**Reply by rasco** [reply]
agree.
one potential complication is that it would have to check all matching routes to see if there is any data.

[product-slug].tsx -> Next.js would try to get the props. if "notFound" is returned, go to the next matching file
[category-slug].tsx -> it tries to fetch the props again, and this time it gets them, and renders the page

**Reply by caomicc** [reply]
Agreed as well, running into this issue with our CMS integration as we have different types of pages + we are trying to keep URLs the same In migration

**Reply by Jacob-Daniel** [reply]
Unless you are going to enforce unique naming convention across 'product' and 'category' for example (undesired) there is nothing preventing naming conflicts.

**Comment by NafiurSiddiqui** [comment]
Alright! So here is the catch! You can do this by using [multiple catch-all segment](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes) by nextJs which @2stash and @popeating  already mentioned.

NextJS do not provide straightforward solution out-of-the-box for this, so here is how you do it.

1. `pages/blog/[...slug].ts` is the root. you can have it straight inside the `pages` root too.
2. have all your X, Y, Z data fetched inside the `getStaticPaths`. 
3. In the return statement of `getStaticPaths`, 

`
return {
   paths: [...X, ...Y, ...Z].map( (page) => ({ params: { slug: page.uri }),
   .....
   }
`
**NOTE** that you must return the `slug` as an array here for this to work. For my case, I splitted up the URI from the first and last index by '/' which returns an array. See what works for you but MUST return an array for this to work.

4. Now get access to this `params` via `context.params.slug` which is available to the `getStaticProps` as `arg`. Now you should be able to dynamically fetch the data based on X, Y, Z properties, pages, posts, whatever.

I used `graphQl` for this for a WP backend and it works fine. It was a massive pain in the bum but eventually, it worked fine.

**Reply by Jacob-Daniel** [reply]
This does not address the question of 'multiple dynamic routes at the same path'.
